### PR TITLE
custom sources per dtag in twap and medianizer combinators

### PR DESCRIPTION
**Pull request by dmfxyz**
left todo

- [x] new tests for TWAP with a ttag
- [x] modify medianizer
- [x] medianizer tests

### CODE REVIEW

File: src/combinators/TWAP.sol [hunk 1]
```diff
@@ -42,6 +43,10 @@ contract TWAP is Ward {
             first.tally = 0;
         }
         configs[tag] = _config;
+
+        //  if targettag is empty, just use same tag
+        if (_config.ttag == bytes32(0)) configs[tag].ttag = tag;
```
Comment (kbrav) [inline src/combinators/TWAP.sol:48]: don't think making it optional adds much value, better to make the user think about the sources of their data

File: src/combinators/TWAP.sol [hunk 2]
```diff
@@ -82,7 +87,8 @@ contract TWAP is Ward {
         } else {
             ttl += config.ttl;
         }
-        fb.push(tag, bytes32((nexttally - pseudotally) / config.range), ttl);
+
+        fb.push(config.ttag, bytes32((nexttally - pseudotally) / config.range), ttl);
```
Comment (kbrav) [inline src/combinators/TWAP.sol:91]: the Config struct has a source address, it should also have a source tag.  more useful to be able to change the sources of twap data than to change the destination.  also we might want multiple twap configs per source

File: src/Medianizer.sol [hunk 3]
```diff
@@ -72,6 +78,7 @@ contract Medianizer {
             median = data[(count - 1) / 2];
         }
 
+        if (tags[tag] != bytes32(0)) tag = tags[tag];
```
Comment (kbrav) [inline src/Medianizer.sol:81]: same thing, let's use source tags.  multiple dests per source vs multiple sources per dest, the idea is to poke the dest

Comment (kbrav) [reply src/Medianizer.sol:81]: yeah also here, we want to have (source, tag)[] => tag.  so maybe one medianized value can be from receiver, one can be from divider, etc

File: src/combinators/TWAP.sol (hunk 1 above)
Comment (dmfxyz) [reply src/combinators/TWAP.sol]: resolved

File: src/combinators/TWAP.sol (hunk 2 above)
Comment (dmfxyz) [reply src/combinators/TWAP.sol]: See [twap map config by dttag, set src tag in config](https://github.com/feedbase/feedbase/pull/97/commits/0fe30c4567a8cf3a9d1be8b743426c13301d88dc)

**Comment by dmfxyz** [comment]
@kbrav [58ed16a](https://github.com/feedbase/feedbase/pull/97/commits/58ed16a854729a07a310ddc0253401296afb8b51) can you take a look before I re-wire all the medianizer tests? Want to make sure this is the intention.

File: src/Medianizer.sol [hunk 4]
```diff
@@ -7,11 +7,13 @@ import './Feedbase.sol';
 contract Medianizer {
     error ErrOwner();
     error ErrQuorum();
+    error ErrSourceTag();
 
     address   public owner;
     address[] public sources;
     Feedbase  public feedbase;
     uint256   public quorum = 1;
+    mapping(bytes32 dtag => mapping (address source => bytes32 tag)) public tags;
```
Comment (kbrav) [inline src/Medianizer.sol:16]: it should be an array of (src, tag) pairs

**Review by kbrav — APPROVED** [review]

File: src/Medianizer.sol [hunk 5]
```diff
@@ -7,11 +7,18 @@ import './Feedbase.sol';
 contract Medianizer {
     error ErrOwner();
     error ErrQuorum();
+    error ErrSourceTag();
```
Comment (kbrav) [inline src/Medianizer.sol:10]: unused?

Comment (dmfxyz) [reply src/Medianizer.sol]: removed

**Pull request merged**